feat(sdk): Implement remove_event_handler

docweirdo opened the pull request:
This PR implements the ability to remove a previously registered event handler.

To remove the event handler, the struct `EventHandlerWrapper` containing the actual handler and an `EventHandlerHandle` is saved in the HashMap for all event handlers instead of event handlers directly. Client.register_event_handler was amended to return an instance of `EventHandlerHandle` associated with the registered handler, which can later be used to remove the event handler with `Client.remove_event_handler`. The Handle contains a uniquely generated ID and the event ID which the event handler is interested in, so the event handler can be removed only with the handle.

This is a breaking change since it removes the method chaining functionality of `Client.register_event_handler`, which is a conscious decision made after consulting with @jplatte.

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -437,48 +439,77 @@ impl Client {
+        let event_id = handle.1;
+        let mut empty = false;
+
+        if let Some(v) = self.inner.event_handlers.write().await.get_mut(&event_id) {
+            v.retain(|e| -> bool { e.handle.0.ne(&handle.0) });
+
+            empty = v.is_empty();
+        }
+
+        if empty {
+            self.inner.event_handlers.write().await.remove(&event_id);
+        }
```
Please only lock the event handler `RwLock` once here. Also, instead of the separate `empty` variable, you can just have a `if v.is_empty()` branch inside the `if let`.

jplatte reviewed (commented):
Since every event handler function is a non-zero-sized `Box`, instead of adding the extra wrapper and attaching a UUID, I think we could just use the address of the handler function as the identity.
I like that the event kind / type are also part of the handle though, so we don't need to iterate all event handlers.

Finally, can you also make `register_event_handler_context` return `Result<()>` instead of `Result<&Self>`? I don't think it makes sense to chain that if the event handler registration can't be chained.

docweirdo commented:
I implemented your suggestions, added some documentation and testing and also implemented an earlier suggestion,  enabling the event handler to remove itself.

I think I forgot to remove the UUID dependency, I will change that with the next commit.

jplatte reviewed (commented):
So, after suggesting this scheme of comparing addresses I unfortunately found a way in which this isn't quite robust enough: If a user was to keep the handle to an event handler after removing that handler, they could potentially remove a different handler with it if the allocator decided to later reuse the same address for that different handler.

Still, I would prefer not going back to the `Uuid` solution, the extra dependency doesn't seem needed. Can you instead add an `AtomicU64` counter to the client and use `.fetch_add(1)` on it to obtain a token that can be used to identify the handler?

docweirdo commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -437,48 +441,121 @@ impl Client {
+    /// # use futures::executor::block_on;
+    /// # use url::Url;
+    /// # use tokio::sync::mpsc;
+    /// # use matrix_sdk_test::{
+    /// #   EphemeralTestEvent, EventBuilder, StateTestEvent, TimelineTestEvent, JoinedRoomBuilder
+    /// # };
```
Yes, I tried to implement an example with the matrix_sdk_test crate but could not call `process_sync(response)` on the mock response as its a private function. I forgot to remove those lines

docweirdo commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -388,23 +393,22 @@ impl Client {
-    ///     .await
-    ///     .register_event_handler(
-    ///         |ev: SyncRoomMessageEvent, room: Room, encryption_info: Option<EncryptionInfo>| {
+    ///     .await;
```
This is actually how it was formatted outside of the docs by the formatter, except that a newline was inserted after each mention of `client` .

jplatte replied:
Right, in that case please do the same here. I find it pretty weird when the closing parenthesis has different indentation from the line that opens the matching opening parenthesis.

docweirdo replied:
I should have done that with the latest commit, right? 

It also seems counter intuitive to me but I'd trust the formatter over my whims any day.

docweirdo commented:
I implemented an `event_handler_counter : AtomicU64` for `Client` which is used to identify event handlers inside the respective `EventHandlerHandle`. Fixed up formatting and the documentation.

I considered using the event handler counter inside the `remove_event_handler` unit test as additional verification, but I would either have to make the field public to the crate or implement a getter function, which I believe would render the atomicity of the counter useless and the test should ensure correctness as is.

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -101,8 +108,8 @@ const DEFAULT_UPLOAD_SPEED: u64 = 125_000;
 type EventHandlerFut = Pin<Box<dyn Future<Output = ()> + Send>>;
-type EventHandlerFn = Box<dyn Fn(EventHandlerData<'_>) -> EventHandlerFut + Send + Sync>;
-type EventHandlerMap = BTreeMap<(EventKind, &'static str), Vec<EventHandlerFn>>;
+pub(crate) type EventHandlerFn = dyn Fn(EventHandlerData<'_>) -> EventHandlerFut + Send + Sync;
```
Is the addition of `pub(crate)` still needed with the current revision of this PR?

docweirdo replied:
Yes, because the type is used in `Client.register_event_handler` to annotate the content of the Box which is passed to `EventHandlerWrapper`.

jplatte replied:
But `register_event_handler` is defined in the same module.

docweirdo replied:
Right, sorry, I mixed it up. It is used to define the type of `EventHandlerWrapper.handler_fn` in the event_handler module.

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -358,6 +371,8 @@ impl Client {
     /// * [`Room`][room::Room] is only available for room-specific events, i.e.
     ///   not for events like global account data events or presence events
     ///
+    /// Returns an [`EventHandlerHandle`].
```
This doesn't add anything, it should either be removed or extended with something like "that can be used to remove the event handler again".

docweirdo replied:
fair point. I think the documentation on `EventHandlerHandle` already states the latter, so I will remove this sentence entirely.

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -388,22 +403,25 @@ impl Client {
+    /// client
     ///     .register_event_handler(
-    ///         |ev: SyncRoomMessageEvent, room: Room, encryption_info: Option<EncryptionInfo>| {
+    ///          |ev: SyncRoomMessageEvent, room: Room, encryption_info: Option<EncryptionInfo>| {
```
One space too much

```suggestion
    ///         |ev: SyncRoomMessageEvent, room: Room, encryption_info: Option<EncryptionInfo>| {
```

docweirdo replied:
good catch!

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -437,48 +455,115 @@ impl Client {
+            .await
+            .entry(H::ID)
+            .or_default()
+            .push(EventHandlerWrapper { handler_function, handle: handle.clone() });
```
I don't see why we should save the "full" handle here, just the `handle_id` should be sufficient.

docweirdo replied:
I need the `EventHandlerHandle` each time the handler is called in `fn handle_sync_events_wrapped_with` to pass it to `EventHandlerData`. I could of course create a new instance of `EventHandlerHandle` each time the handler is called, but that is less efficient in my eyes.

jplatte replied:
I'm pretty sure it has almost no cost. I would prefer if it was done that way, but it might also be harder. If it isn't easy / you don't want to do it, I'm okay merging this PR without that and revisiting it later.

docweirdo replied:
I just tried to change it, but to create a new `EventHandlerHandle` within `fn handle_sync_events_wrapped_with` I need the variable `ev_type` which, since it is inferred from `raw_event`, has a lifetime of `'a` instead of `'static'. 

I will leave it as is since that is alright with you.

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -437,48 +455,115 @@ impl Client {
+    /// `handle` - The [`EventHandlerHandle`] that is returned when
+    /// registering the event handler with [`Client::register_event_handler`].
+    ///
+    /// # Examples
```
This example doesn't actually use `remove_event_handler`.

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -511,22 +596,20 @@ impl Client {
-    ///     .register_event_handler_context(my_gui_handle.clone())
-    ///     .register_event_handler(
+    /// client.register_event_handler_context(my_gui_handle.clone());
+    /// client.register_event_handler(
```
```suggestion
    /// client
    ///     .register_event_handler(
```

jplatte commented on crates/matrix-sdk/src/event_handler.rs:
```diff
@@ -83,6 +84,25 @@ pub trait SyncEvent {
 }
 
+pub(crate) struct EventHandlerWrapper {
+    pub handler_function: Box<EventHandlerFn>,
```
```suggestion
    pub handler_fn: Box<EventHandlerFn>,
```

jplatte commented on crates/matrix-sdk/src/event_handler.rs:
```diff
@@ -83,6 +84,25 @@ pub trait SyncEvent {
+    pub handle: EventHandlerHandle,
+}
+
+#[derive(Debug, Clone)]
+/// Handle to remove a registered event handler by passing it to
+/// [`Client::remove_event_handler`].
```
```suggestion
/// Handle to remove a registered event handler by passing it to
/// [`Client::remove_event_handler`].
#[derive(Clone, Copy, Debug)]
```

jplatte commented on crates/matrix-sdk/src/event_handler.rs:
```diff
@@ -83,6 +84,25 @@ pub trait SyncEvent {
+/// Handle to remove a registered event handler by passing it to
+/// [`Client::remove_event_handler`].
+pub struct EventHandlerHandle {
+    pub(crate) id: (EventKind, &'static str),
+    pub(crate) addr: u64,
```
The `addr` name no longer makes sense. How about this (the current `ID` tuple is removed in #878 anyways):

```suggestion
    pub(crate) ev_kind: EventKind,
    pub(crate) ev_type: &'static str,
    pub(crate) id: u64,
```

docweirdo replied:
Since we need a tuple of `ev_kind` *and* `ev_type` and never really only one of them, I will save both as a tuple called `ev_id`, whereas the other id will be called `handler_id` for it identifies the handler and not the handle.

jplatte replied:
Do we really need them to be "tupled" together?

docweirdo replied:
I can deconstruct them, then put them each in a field and put them together afterwards, but I don't see why that's more useful?

jplatte replied:
Hm, okay. I guess because the tuple is currently used as the map key, it's okay. I intend to change that soon, but can "un-tuple" the fields then.

jplatte commented:
Side note: With #878 merged, you'll now have to rebase. The merge conflicts should be minimal though.

docweirdo commented:
I noticed that using `Atomic64` for an `event_handler_counter` and a `u64` as the id for an event handler poses the same challenge as using the memory address of the handler for identification, namely that when holding onto a clone of `EventHandlerHandle` (since the remove function consumes the handle) and passing it to `remove_event_handler` at a later point in time, it might deregister an unrelated handler. 

I would either label the possibility of that happening as very low or at least call it self inflicted and leave it as is or switch back to the addresses because it removes an unnecessary field from `Client`, or go back to using `UUID`s.

jplatte commented:
I don't understand, how would it be possible for an "old" handle to match a handler that was created later? Are you considering `u64` wrap-around, or something else?

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -437,48 +453,113 @@ impl Client {
+            }
+        }
+
+        self.inner.event_handler_counter.fetch_sub(1, SeqCst);
```
I guess this is what you're talking about regarding handle uniqueness. I have an easy fix for you 😉 

```suggestion
```

docweirdo replied:
alrighty :D

docweirdo replied:
this will limit us to (2^64)-1 event handlers

docweirdo commented:
I've rebased on `main`.

jplatte commented on crates/matrix-sdk/src/event_handler.rs:
```diff
@@ -86,6 +87,25 @@ pub trait SyncEvent {
+
+/// Handle to remove a registered event handler by passing it to
+/// [`Client::remove_event_handler`].
+#[derive(Debug, Clone)]
```
```suggestion
#[derive(Clone, Copy, Debug)]
```

jplatte commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -446,39 +461,103 @@ impl Client {
+        let mut event_handlers = self.inner.event_handlers.write().await;
+
+        if let Some(v) = event_handlers.get_mut(&handle.ev_id) {
+            v.retain(|e| -> bool { e.handle.handler_id.ne(&handle.handler_id) });
```
```suggestion
            v.retain(|e| e.handle.handler_id != handle.handler_id);
```
